src/compose/images.ts

```diff
@@ -52,584 +47,599 @@ type NormalisedDockerImage = Docker.ImageInfo & {
-					id = await this.fetchDelta(image, opts, onProgress, serviceName);
-				} else {
-					id = await this.fetchImage(image, opts, onProgress);
-				}
+		try {
+			let id;
+			if (opts.delta && (opts as DeltaFetchOptions).deltaSource != null) {
+				id = await fetchDelta(image, opts, onProgress, serviceName);
+			} else {
+				id = await fetchImage(image, opts, onProgress);
+			}
 
-				await db.models('image').update({ dockerImageId: id }).where(image);
-
-				logger.logSystemEvent(LogTypes.downloadImageSuccess, { image });
-				success = true;
-				delete this.imageFetchFailures[image.name];
-				delete this.imageFetchLastFailureTime[image.name];
-			} catch (err) {
-				if (err instanceof DeltaStillProcessingError) {
-					// If this is a delta image pull, and the delta still hasn't finished generating,
-					// don't show a failure message, and instead just inform the user that it's remotely
-					// processing
-					logger.logSystemEvent(LogTypes.deltaStillProcessingError, {});
-				} else {
-					this.addImageFailure(image.name);
-					logger.logSystemEvent(LogTypes.downloadImageError, {
-						image,
-						error: err,
-					});
-				}
-				success = false;
+			await db.models('image').update({ dockerImageId: id }).where(image);
+
+			logger.logSystemEvent(LogTypes.downloadImageSuccess, { image });
+			success = true;
+			delete imageFetchFailures[image.name];
+			delete imageFetchLastFailureTime[image.name];
+		} catch (err) {
+			if (err instanceof DeltaStillProcessingError) {
+				// If this is a delta image pull, and the delta still hasn't finished generating,
+				// don't show a failure message, and instead just inform the user that it's remotely
+				// processing
+				logger.logSystemEvent(LogTypes.deltaStillProcessingError, {});
+			} else {
+				addImageFailure(image.name);
+				logger.logSystemEvent(LogTypes.downloadImageError, {
+					image,
+					error: err,
+				});
 			}
+			success = false;
 		}
-
-		this.reportChange(image.imageId);
-		onFinish(success);
-		return null;
 	}
 
-	public async remove(image: Image): Promise<void> {
-		try {
-			await this.removeImageIfNotNeeded(image);
-		} catch (e) {
-			logger.logSystemEvent(LogTypes.deleteImageError, {
-				image,
-				error: e,
-			});
-			throw e;
-		}
-	}
+	reportChange(image.imageId);
+	onFinish(success);
+	return null;
+}
 
-	public async getByDockerId(id: string): Promise<Image> {
-		return await db.models('image').where({ dockerImageId: id }).first();
+export async function remove(image: Image): Promise<void> {
+	try {
+		await removeImageIfNotNeeded(image);
+	} catch (e) {
+		logger.logSystemEvent(LogTypes.deleteImageError, {
+			image,
+			error: e,
+		});
+		throw e;
 	}
+}
 
-	public async removeByDockerId(id: string): Promise<void> {
-		const image = await this.getByDockerId(id);
-		await this.remove(image);
-	}
+export async function getByDockerId(id: string): Promise<Image> {
+	return await db.models('image').where({ dockerImageId: id }).first();
```

Comment:
We can remove the await and just return. In examples such as: https://github.com/balena-io/balena-supervisor/blob/abeb33ae1fd9b14e014c267df9c903368bbe064c/src/compose/images.ts#L413-L415 we have return await so the rejection gets caught in the try rather then returned. There are a few instances in the PR we can remove await in the return.